connector/components/locker.py

```diff
@@ -0,0 +1,69 @@
+# Copyright 2018 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html)
+
+import logging
+import psycopg2
+
+from odoo.addons.component.core import Component
+from odoo.addons.connector.exception import RetryableJobError
+
+_logger = logging.getLogger(__name__)
+
+
+class RecordLocker(Component):
+    """Component allowing to lock record(s) for the current transaction
+
+    Example of usage::
+
+        self.component('record.locker').lock(self.records)
+
+    See the definition of :meth:`~lock` for details.
+    """
+
+    _name = "base.record.locker"
+    _inherit = ["base.connector"]
+    _usage = "record.locker"
+
+    def lock(self, records, seconds=None, ignore_retry=True):
+        """Lock the records.
+
+        Lock the record so we are sure that only one job is running for this
+        record(s) if concurrent jobs have to run a job for the same record(s).
+        When concurrent jobs try to work on the same record(s), the first one
+        will lock and proceed, the others will fail to acquire it and will be
+        retried later
+        (:exc:`~odoo.addons.queue_job.exception.RetryableJobError` is raised).
+
+        The lock is using a ``FOR UPDATE NOWAIT`` so any concurrent transaction
+        trying FOR UPDATE/UPDATE will be rejected until the current transaction
+        is committed or rollbacked.
+
+        A classical use case for this is to prevent concurrent exports.
+
+        The following parameters are forwarded to the exception
+        :exc:`~odoo.addons.queue_job.exception.RetryableJobError`
+
+        :param seconds: In case of retry because the lock cannot be acquired,
+                        in how many seconds it must be retried. If not set,
+                        the queue_job configuration is used.
+        :param ignore_retry: If True, the retry counter of the job will not be
+                             increased.
+        """
+        sql = ("SELECT id FROM %s WHERE ID IN %%s FOR UPDATE NOWAIT" %
+               self.model._table)
+        try:
+            self.env.cr.execute(sql, (tuple(records.ids),),
+                                log_exceptions=False)
+        except psycopg2.OperationalError:
+            _logger.info(
+                "A concurrent job is already working on the same "
+                "record (%s with one id in %s). Job delayed later.",
+                self.model._name,
+                tuple(records.ids),
+            )
+            raise RetryableJobError(
+                "A concurrent job is already working on the same record "
+                "(%s with one id in %s). The job will be retried later."
+                % (self.model._name, tuple(records.ids),),
```

Comment:
does this string interpolation work? Shouldn't you pass `str(records.ids)`?

Reply:
No, this works. The string substitution implies the cast to a string type.

Reply:
nope, this works, by chance only when you have 1 var x 1 slot, meaning:
```bash
>>> 'Interpolate this please %s' % (tuple([1,2,3]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: not all arguments converted during string formatting
>>> 'Interpolate this please %s' % (tuple([1,]))
'Interpolate this please 1'
```

Reply:
@simahawk you are not passing a tuple for the interpolation, the code in the PR does:

```python
>>> 'Interpolate this please %s' % (tuple([1,2,3]),)
'Interpolate this please (1, 2, 3)'
>>> 'Interpolate this please %s' % (tuple([1,]),)
'Interpolate this please (1,)'
```

Reply:
Yes, but the author shows he is aware of this. Note the extra comment after `tuple(record.ids)`

Reply:
hahahaha I definitely need some rest... tricked by sh*tty comma :sweat_smile: 
sorry for the noise guys :stuck_out_tongue: